### PR TITLE
ci: Add HDF5 to a windows build

### PR DESCRIPTION
**Pull request by scottwittenburg**
Turns on HDF5 for windows builds using the vs2022 compiler.

Fixes: #3662

### CODE REVIEW

**Comment by scottwittenburg** [comment]
The [build](https://github.com/ornladios/ADIOS2/actions/runs/5742192428/job/15563855190?pr=3732) turned up warnings (which were treated as errors) like this one:

```
  D:\a\ADIOS2\ADIOS2\source\source\adios2\engine\hdf5\HDF5ReaderP.cpp(266,1): error C2220: the following warning is treated as an error [D:\a\ADIOS2\ADIOS2\win2022-vs2022-ompi\source\adios2\adios2_hdf5.vcxproj]
  ...
  D:\a\ADIOS2\ADIOS2\source\source\adios2\engine\hdf5\HDF5ReaderP.cpp(266,1): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data [D:\a\ADIOS2\ADIOS2\win2022-vs2022-ompi\source\adios2\adios2_hdf5.vcxproj]
  D:\a\ADIOS2\ADIOS2\source\source\adios2\engine\hdf5\HDF5ReaderP.cpp(186,1): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data [D:\a\ADIOS2\ADIOS2\win2022-vs2022-ompi\source\adios2\adios2_hdf5.vcxproj]
```

Looking at the code, I wonder why in `VariableBase` the `m_StepStart` is defined as `size_t`, while the `HDF5Common` api thinks of steps as `int`?  Should steps be considered as `size_t` across the entire codebase?  Or should we just continue with the implicit conversion but do some runtime check/assert first?  Or something else?

@vicentebolea

**Comment by vicentebolea** [comment]
> Looking at the code, I wonder why in VariableBase the m_StepStart is defined as size_t, while the HDF5Common api thinks of steps as int? Should steps be considered as size_t across the entire codebase? Or should we just continue with the implicit conversion but do some runtime check/assert first? Or something else?

There might some reason to make it interoperable with HDF5, @guj more expertise in those modules.

**Comment by guj** [comment]
> > Looking at the code, I wonder why in VariableBase the m_StepStart is defined as size_t, while the HDF5Common api thinks of steps as int? Should steps be considered as size_t across the entire codebase? Or should we just continue with the implicit conversion but do some runtime check/assert first? Or something else?
> 
> There reason might be to make it interoperable with HDF5, @guj more expertise in those modules.

Sorry I just saw this. Scott is right, steps should be size_t. I will correct it in HDF5Common.* in another pull request.

**Comment by scottwittenburg** [comment]
> Sorry I just saw this. Scott is right, steps should be size_t. I will correct it in HDF5Common.* in another pull request.

Sounds good, thanks @guj!

**Comment by guj** [comment]
> > Sorry I just saw this. Scott is right, steps should be size_t. I will correct it in HDF5Common.* in another pull request.
> 
> Sounds good, thanks @guj!

**Comment by scottwittenburg** [comment]
@guj I think we still want this, since it will allow testing HDF5+windows in CI.  I'm re-opening and will rebase on the latest `master`, which should allow verifying the fixes you merged last week.

**Comment by scottwittenburg** [comment]
@vicentebolea I can reproduce the 6 errors in the Windows/VS2022 builds locally, and I added the latest WIP commit to try to see what's happening in CI when the test passes vs fails.  I'm guessing the spack-installed hdf5 in our ubuntu images is missing the extra `COMPONENT=HL` I had to add for debugging, so those all failed to configure in the latest builds.  

But still I can see the debug output from the failed tests as well as from at least one passing.  The test failure I instrumented with extra debug output is happening at this [line](https://github.com/ornladios/ADIOS2/pull/3732/files#diff-5735856915e893d6f3fc839871d56b9000fb54a23c5c0d8e1529308f6094f4baR638).  And what's weird to me is that the `h5Type` variable which is supposed to equal `H5T_NATIVE_LONG` has the same description (`H5T_STD_I64LE`) in both the [failed](https://open.cdash.org/test/1198057260) and [passing](https://open.cdash.org/test/1198064331) tests. 

The HDF5 docs have a table (a little ways down [here](https://docs.hdfgroup.org/hdf5/develop/_l_b_datatypes.html)) which says that the "native" type `H5T_NATIVE_LONG` corresponds to, among a few other "standard" types, `H5T_STD_I64LE`.  So I don't understand why that isn't holding true for Windows/VS2022.

If you or @guj have any thoughts about what could be going on, please let me know.  Thanks!

**Comment by guj** [comment]
> @vicentebolea I can reproduce the 6 errors in the Windows/VS2022 builds locally, and I added the latest WIP commit to try to see what's happening in CI when the test passes vs fails. I'm guessing the spack-installed hdf5 in our ubuntu images is missing the extra `COMPONENT=HL` I had to add for debugging, so those all failed to configure in the latest builds.
> 
> But still I can see the debug output from the failed tests as well as from at least one passing. The test failure I instrumented with extra debug output is happening at this [line](https://github.com/ornladios/ADIOS2/pull/3732/files#diff-5735856915e893d6f3fc839871d56b9000fb54a23c5c0d8e1529308f6094f4baR638). And what's weird to me is that the `h5Type` variable which is supposed to equal `H5T_NATIVE_LONG` has the same description (`H5T_STD_I64LE`) in both the [failed](https://open.cdash.org/test/1198057260) and [passing](https://open.cdash.org/test/1198064331) tests.
> 
> The HDF5 docs have a table (a little ways down [here](https://docs.hdfgroup.org/hdf5/develop/_l_b_datatypes.html)) which says that the "native" type `H5T_NATIVE_LONG` corresponds to, among a few other "standard" types, `H5T_STD_I64LE`. So I don't understand why that isn't holding true for Windows/VS2022.
> 
> If you or @guj have any thoughts about what could be going on, please let me know. Thanks!

Thanks @scottwittenburg for checking on windows!!
Looking from these two docs:
https://learn.microsoft.com/en-us/windows/win32/winprog/windows-data-types
https://learn.microsoft.com/en-us/cpp/c-runtime-library/standard-types?view=msvc-170
My guess is,  64 bit integer type H5T_STD_I64LE is treated as long long.   
so asserting with H5T_NATIVE_LLONG might pass for windows.  

similarly for the later assertion of type H5T_NATIVE_ULONG ..

**Comment by scottwittenburg** [comment]
> so asserting with H5T_NATIVE_LLONG might pass for windows, similarly for the later assertion of type H5T_NATIVE_ULONG ..

You were correct @guj, thank you for your help :grin:  

In [2e37421](https://github.com/ornladios/ADIOS2/pull/3732/commits/2e37421e23e3211a5dcffa1672b870b82fa54fe6) I've taken a stab at what you suggested, and it has resolved the remaining windows+hdf5 failures.  If you get a chance, can you take a look and make sure that fix looks good to you?  Thanks again!

**Comment by guj** [comment]
> > so asserting with H5T_NATIVE_LLONG might pass for windows, similarly for the later assertion of type H5T_NATIVE_ULONG ..
> 
> You were correct @guj, thank you for your help 😁
> 
> In [2e37421](https://github.com/ornladios/ADIOS2/pull/3732/commits/2e37421e23e3211a5dcffa1672b870b82fa54fe6) I've taken a stab at what you suggested, and it has resolved the remaining windows+hdf5 failures. If you get a chance, can you take a look and make sure that fix looks good to you? Thanks again!

Looks good to me! Thanks @scottwittenburg !

File: scripts/ci/cmake/ci-win2022-vs2022-ompi.cmake [hunk 1]
```diff
@@ -3,9 +3,15 @@ set(ENV{CXX} cl)
 set(ENV{CFLAGS} /WX)
 set(ENV{CXXFLAGS} /WX)
 
+# Tests need to find hdf5.dll
+set(ENV{PATH} "$ENV{PATH};C:/hdf5/HDF5-1.14.2.1-win64/bin")
```
Comment (vicentebolea) [inline scripts/ci/cmake/ci-win2022-vs2022-ompi.cmake:7]: :+1: I am happy that it worked :sunglasses:

File: examples/heatTransfer/write/IO_hdf5_a.cpp [hunk 2]
```diff
@@ -49,7 +49,7 @@ void IO::write(int step, const HeatTransfer &ht, const Settings &s, MPI_Comm com
     double total_time = MPI_Wtime() - time_start;
     uint64_t adios_totalsize = 2 * sizeof(int) + 2 * s.ndx * s.ndy * sizeof(double);
     uint64_t sizeMB = adios_totalsize * s.nproc / 1024 / 1024 / 1024; // size in MB
-    uint64_t mbs = sizeMB / total_time;
+    double mbs = sizeMB / total_time;
```
Comment (vicentebolea) [inline examples/heatTransfer/write/IO_hdf5_a.cpp:52]:
```suggestion
    double mbs = sizeMB / static_cast<double>(total_time);
```

C/C++ casting rules, I made an example here https://godbolt.org/z/rvWGzPKoj

**Review by vicentebolea — APPROVED** [review]
Looks Good. I proposed a change in one the examples, apply the suggestion if you think that it makes sense and I will approve again. 

In any case already approved.

File: examples/heatTransfer/write/IO_hdf5_a.cpp (hunk 2 above)
Comment (scottwittenburg) [reply examples/heatTransfer/write/IO_hdf5_a.cpp:52]: I could be misinterpreting your suggestion, but isn't `total_time` already of type `double`?  Though you're correct about division when both operands are of type `int`, I think as long as one operand is of type `double`, the other will be promoted before the operation.  Let me know if I misunderstood though.

**Comment by scottwittenburg** [comment]
@vicentebolea The `crusher-kokkos-hip` build failed with:

```
slurmstepd: error: *** JOB 367818 ON crusher189 CANCELLED AT 2023-08-31T15:37:19 DUE TO TIME LIMIT ***
```

Have you seen that before?  Should I just retry that job?

**Comment by vicentebolea** [comment]
> @vicentebolea The `crusher-kokkos-hip` build failed with:
> 
> ```
> slurmstepd: error: *** JOB 367818 ON crusher189 CANCELLED AT 2023-08-31T15:37:19 DUE TO TIME LIMIT ***
> ```
> 
> Have you seen that before?  Should I just retry that job?

Hiccups at the olcf ci, i will retry the job a bit later

**Comment by vicentebolea** [comment]
@scottwittenburg CI passing

**Comment by vicentebolea** [comment]
@scottwittenburg feel free to merge if you see fit, approved :)

**Pull request merged**